cpp/src/arrow/compute/kernels/vector_selection_filter_internal.cc

```diff
@@ -444,196 +518,244 @@ Status PrimitiveFilterExec(KernelContext* ctx, const ExecSpan& batch, ExecResult
-      in_position += filter_block.length;
-    } else if (filter_valid_block.AllSet()) {
-      // Simpler path: no filter values are null
-      if (filter_block.AllSet()) {
-        // Fastest path: filter values are all true and not null
-        if (values_valid_block.AllSet()) {
-          // The values aren't null either
-          bit_util::SetBitsTo(out_is_valid, out_position, filter_block.length, true);
-
-          // Bulk-append raw data
-          offset_type block_data_bytes =
-              (raw_offsets[in_position + filter_block.length] - raw_offsets[in_position]);
-          APPEND_RAW_DATA(raw_data + raw_offsets[in_position], block_data_bytes);
-          // Append offsets
-          for (int64_t i = 0; i < filter_block.length; ++i, ++in_position) {
-            offset_builder.UnsafeAppend(offset);
-            offset += raw_offsets[in_position + 1] - raw_offsets[in_position];
-          }
-          out_position += filter_block.length;
-        } else {
-          // Some of the values in this block are null
-          for (int64_t i = 0; i < filter_block.length;
-               ++i, ++in_position, ++out_position) {
-            offset_builder.UnsafeAppend(offset);
-            if (bit_util::GetBit(values_is_valid, values_offset + in_position)) {
-              bit_util::SetBit(out_is_valid, out_position);
-              APPEND_SINGLE_VALUE();
-            }
+  if (is_ree_filter) {
+    auto emit_segment = [&](int64_t position, int64_t segment_length, bool filter_valid) {
+      in_position = position;
+      if (filter_valid) {
+        // Filter values are all true and not null
+        // Some of the values in the block may be null
+        for (int64_t i = 0; i < segment_length; ++i, ++in_position, ++out_position) {
+          offset_builder.UnsafeAppend(offset);
+          if (bit_util::GetBit(values_is_valid, values_offset + in_position)) {
+            bit_util::SetBit(out_is_valid, out_offset + out_position);
+            APPEND_SINGLE_VALUE();
           }
         }
-      } else {  // !filter_block.AllSet()
-        // Some of the filter values are false, but all not null
-        if (values_valid_block.AllSet()) {
-          // All the values are not-null, so we can skip null checking for
-          // them
-          for (int64_t i = 0; i < filter_block.length; ++i, ++in_position) {
-            if (bit_util::GetBit(filter_data, filter_offset + in_position)) {
+      } else {
+        offset_builder.UnsafeAppend(segment_length, offset);
+        out_position += segment_length;
+      }
+      return Status::OK();
+    };
+    Status status;
+    VisitPlainxREEFilterOutputSegments(
+        filter, true, null_selection,
+        [&status, emit_segment = std::move(emit_segment)](
+            int64_t position, int64_t segment_length, bool filter_valid) {
+          status = emit_segment(position, segment_length, filter_valid);
+          return status.ok();
+        });
+    RETURN_NOT_OK(std::move(status));
```

Comment:
The `std::move` is a bit pedantic here IMHO...

Reply:
`GenericToStatus` takes an r-value and that avoids code-bloat -- no need to generate memory allocation code to copy the `Status` string. This is multiplied by the number of template instances.

```cpp
  #define ARROW_RETURN_NOT_OK(status)                                   \
    do {                                                                \
      ::arrow::Status __s = ::arrow::internal::GenericToStatus(status); \
      ARROW_RETURN_IF_(!__s.ok(), __s, ARROW_STRINGIFY(status));        \
    } while (false)
```

Reply:
```cpp
Status::Status(const Status& s)
      : state_((s.state_ == NULLPTR) ? NULLPTR : new State(*s.state_)) {}
```

vs

```cpp
  Status::Status(Status&& s) noexcept : state_(s.state_) { s.state_ = NULLPTR; }
```

being inlined.

Reply:
Which yes, is a pedantic `std::move`, but should I remove it? :D

Reply:
No need to remove it. I'm merely pointing out that it's not necessary to micro-optimize this particular end of function :-)